File: leetcodec++/leetcode279.cpp
//
// Created by 黄敏辉 on 2019-09-18.
//

class Solution {
public:
    int numSquares(int n) {
        vector<int> dp{0};
        while (dp.size() <= n){
            int m = dp.size(), s = INT_MAX;
            for (int i = 1; i*i <= m; i++)
                s = min(s, dp[m - i*i] + 1);
            dp.push_back(s);
        }
        return dp[n];
    }
};

File: python/leetcode279.py

class Solution:
    _dp = [0]
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = self._dp
        while len(dp)<=n:
            dp.append(min([dp[-i*i] for i in range(1, int(len(dp)**0.5+1))])+1)
        return dp[n]